Abrt support

marusak opened the pull request:
This PR implements design 
https://github.com/cockpit-project/cockpit/wiki/Feature:-ABRT

This PR needs upstream Libreport (https://github.com/abrt/libreport), and newest ABRT. All of this will be in Fedora 26. However, if older ABRT or Libreport are installed or are not available at all no changes are visible for users.
If you want to test this PR now, you will need a few thinks:
abrt >= 2.8.0
libreport - must be build from upstream
satyr >= 0.22

It is also possible to use repository that I created in copr.
https://copr.fedorainfracloud.org/coprs/mmarusak/ABRT-for-cockpit-PR/

When everything installed you need to run:
service abrtd start
service abrt-ccpp stop
service abrt-journal-core start
ulimit -c unlimited
setenforce 0

All of these steps are only necessary because behaviour of Libreport is changing quite a bit from Fedora 26 onwards.

Then log in as root. (You need to be a sudo user to read problems. Otherwise none problems will be displayed.)

andreasn commented:
I was able to run this, but I couldn't get anything to show up under Problems.
Is there an easy way to generate a problem that will show up there?

marusak commented:
Yes, there are two easy ways how to do it.
The easiest is to install will-crash. Then you can run for example will-abort or will-segfault.
Or if you don't want to install anything, you can kill sleep for example.
sleep 10000m &
kill -SIGSEGV %%

marusak commented:
I see that 3 tests failed. 
selenium/chrome and selenium/firefox failed due to test50ChangeTabLogs in test/avocado/selenium-base.py. This is my fault that I forgot to update test when I changed buttons in https://github.com/cockpit-project/cockpit/pull/6189/commits/b1bc82b I will fix this test.
However test verify/fedora-25 timed out and I am not sure why.

andreasn commented:
Unfortunately I am still unable to run this on Fedora 25. I tried all the steps above, but no success.
I'll try to ping you on irc when you're around again.

andreasn commented:
Works great in general! Thank you so much for getting this together!
Mostly small styling nits from my side:

The header looks a bit too tall, and the text is a bit too big, compared to other places we use this pattern:
![screenshot from 2017-03-29 15-11-15](https://cloud.githubusercontent.com/assets/209895/24456198/93033ee6-1492-11e7-916d-73cf7b301c1e.png)

Compare it to how it looks on the [Storage page](https://cloud.githubusercontent.com/assets/209895/24456263/c7295b24-1492-11e7-9df3-e0a4f2e87a22.png) and the [react examples page](https://cloud.githubusercontent.com/assets/209895/24456878/af36e8a4-1494-11e7-8348-c31e938e2c31.png)


The column that starts with "reason" starts a lot further down than the column that starts with "Directory". The left column needs to line up with the right one at the top:
![screenshot from 2017-03-29 15-12-17](https://cloud.githubusercontent.com/assets/209895/24456280/d8eb054c-1492-11e7-88fb-cf5271586487.png)


I get an errors along the lines of "Reporting was unsucessful. Try running `reporter-ureport -d /var/spool/abrt/ccpp-2017-03-29-10:55:36-24549`" when I try to Report an issue.
The error bar is probably better placed below Logs >> Entry and above the box with the error, so it doesn't break the layout.
![screenshot from 2017-03-29 15-15-20](https://cloud.githubusercontent.com/assets/209895/24456354/0dae80ba-1493-11e7-902a-0a22f32e395c.png)


Even though I got an error that the reporting failed, the spinner still spins like it's trying to report.


Clicking another tab while the spinner is still going starts displaying the contents of that above the tab area for some reason.
![screenshot from 2017-03-29 15-22-12](https://cloud.githubusercontent.com/assets/209895/24456610/d5711e3c-1493-11e7-83f5-f6638b8ce719.png)


The expander and text is closer to the top than the bottom, and the rows are only expandable if clicking right on the text. I think it would be better if it was possible to expand and collapse by clicking anywhere in the row.
![screenshot from 2017-03-29 15-26-30](https://cloud.githubusercontent.com/assets/209895/24456935/de39b7b2-1494-11e7-8dd7-13ea36505bbc.png)

marusak commented:
Thanks for the review. I addressed all of yours points and they are all fixed now.

andreasn commented:
Thank you for the quick fixes!

I went over it again, and found a few, tiny issues more.

The delete icon is different from what we use in the rest of Cockpit, and what's recommended by Patternfly (see http://www.patternfly.org/styles/icons/). I would suggest to use pficon-remove (we use the name pficon-delete elsewhere in Cockpit it seems, but the style is the same...will fix that in the other places in Cockpit)

I get this dropdown with only one thing in it when I've successfully reported a problem.
![screenshot from 2017-03-31 20-06-16](https://cloud.githubusercontent.com/assets/209895/24563605/4c491dea-164f-11e7-8da0-159ee13a8cc4.png)
I would suggest to make this a link instead, as that communicates the outcome of the action a bit clearer (ie. that it will open a URL in the browser).

There is a bit of space between the tabs. Tried to figure out how it's different from what we use in other places in Cockpit, but haven't been able to nail it down quite.
![screenshot from 2017-03-31 20-10-29](https://cloud.githubusercontent.com/assets/209895/24563779/eac906c4-164f-11e7-9dc7-6d28dcd83430.png)

The entire rows are clickable now. Awesome! The pointer is different depending on where you hover in the row still though. Needs a cursor: pointer; on :hover

Apart from those, everything looks and behaves as expected!

andreasn commented:
> The entire rows are clickable now. Awesome! The pointer is different depending on where you hover in the row still though. Needs a cursor: pointer; on :hover

Realized a better solution is might be to give `width: 100%;` and `display: block;` to the a-element.
http://stackoverflow.com/questions/4861230/make-a-link-have-100-width

@garrett is better at this kind of stuff though. What do you think, Garrett?

marusak commented:
Great points.

> The delete icon is different from what we use in the rest of Cockpit...

Fixed.

> I get this dropdown with only one thing in it when I've successfully reported a problem.

That is on purpose. You can get more than one URL. There can be BZs for example.

>There is a bit of space between the tabs.

https://www.patternfly.org/pattern-library/widgets/#tabs it seems you are non-standard from patternfly. I was not able to find out what you have different from patternfly. I will try more, but if somebody knows what is different I would be happy to learn that.

>Needs a cursor: pointer; on :hover

Done. Only `cursor: pointer` on panel-heading was necessary. I also added grey colour of panel on hover. Reason why I have not used `width: 100%; and display: block; to the a-element` is that it did not cover the whole panel only the middle part (in vertical way).

andreasn commented:
> it seems you are non-standard from patternfly

Indeed!
@garrett said he'll look into this. Can be fixed as a followup (either that this get fixed to be more like the rest of cockpit, or that the rest of cockpit gets fixed to be more like patternfly). So let's leave it like this for now.

> That is on purpose. You can get more than one URL. There can be BZs for example.

Is there any support for that in this PR, or is it for future-proofing?
What other entrys can it have in there?
Will it usually be one, two or more entrys?

The dropdown seems to cause page-resize here.
![screenshot from 2017-04-03 18-07-03](https://cloud.githubusercontent.com/assets/209895/24619355/fe641fe2-1899-11e7-9307-0e5ec42fadc1.png)
One workaround might be to override the `dropdown-menu` class sets. With something like `0` or `auto` might need an important flag too.

stefwalter commented:
Looking good. I noticed you recorded this video: https://www.youtube.com/watch?v=jYm7jt27X04

marusak commented:
@andreasn 
> Is there any support for that in this PR, or is it for future-proofing?
> What other entrys can it have in there?
> Will it usually be one, two or more entrys?

As I was trying to answer your questions I realised that my approach was not ideal. There can be more than one entry. But the entry would be only available if you used other ABRT tools such as gnome-abrt or abrt-cli. When I was doing that, my main idea was that in future when reporting to BZ will be added, that the BZ link could show under that dropdown (Because ABRT does provide all URLs in one array). However I realised that more than 2 links will not be often if at all. And also it would be confusing for users to have two reporting button but results would show under one dropdown. 
There are still some questions how the reporting in future should look like. When we decide to implement that, I will write all usecases and options and will ask you for you option. But that is not in this PR.
In short I dropped the dropdown menu and only 'Reported' clickable string is displayed as in https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/crashes/crashes-journal.png . That also solved the problem with page-resize.

@stefwalter I have. It was mainly for my manager since he wanted to see as this integration is going. Any comments, ideas?

stefwalter commented:
@marusak I'll let @andreasn make the judgement calls on the functionality and design.

But as far as the video. It's great that you recorded a video. We usually require a video or screenshot recorded for new user visible features before they get merged. That way we can include them in the (weekly) release notes.

andreasn commented:
> However I realised that more than 2 links will not be often if at all. And also it would be confusing for users to have two reporting button but results would show under one dropdown.
There are still some questions how the reporting in future should look like. When we decide to implement that, I will write all usecases and options and will ask you for you option. But that is not in this PR.

Makes sense. Thanks!

garrett commented:
I saw the video and heard the distracting background noise. (Was this recorded on an internal microphone on a laptop with high load and fans?) Since I taught myself how to use Kdenlive and Audacity recently, I figured I'd try to put them to use and see if I could help clean up the audio.

After downloading the video from YouTube, I opened it in Kdenlive, split out the audio track, processed the audio to remove the background noise and adjust the levels (somewhat), then merged it all back together in Kdenlive, and exported it. Then, of course, I uploaded it to YouTube.

New video, with cleaned-up audio @ https://www.youtube.com/watch?v=Uw7LIJpfLI0

This gave me some real-world usage of Kdenlive & Audacity before I get the videos from Andreas. It also tested my laptop to make sure it is set up properly to perform all these tasks.

Also, @marusak, I would be happy to give you the modified video file to replace yours, if you wish. :smile:

marusak commented:
@garrett Yes, the noise was due to my laptop overheating and there was nothing I could do in the time of recording the video. I was not happy with the sound quality at all.
I am genuinely surprised how you could edit the noise out. Great work! 
It would be very kind from you if you could send me the edited sound on my email mmarusak@redhat.com.
Thanks:)

mvollmer commented on pkg/lib/journal.js:
```diff
@@ -230,7 +234,34 @@
+                var all_p = [entry['PROBLEM_DIR'], entry['PROBLEM_DUPHASH'], entry['PROBLEM_UUID']];
+                var p;
+                var known = false;
+                for (var i = 0; i < 3; i++) {
```
The "3" should be replaced with `all_p.length`, I'd say.

marusak replied:
Fixed

mvollmer commented on pkg/systemd/logs.js:
```diff
@@ -288,6 +380,391 @@ $(function() {
+
+        for (var i = 0; i < lines.length - 1; i++) {
+            var line = lines[i].split(delimiter);
+            result += '<tr> <td style="test-align: right">' + line[0];
```
`text-align`

mvollmer replied:
We can not use inline styles like this because of the Content-Security-Policy.  Just use `class="text-right"`.

marusak replied:
Fixed

mvollmer commented on pkg/systemd/logs.js:
```diff
@@ -288,6 +380,391 @@ $(function() {
+        for (var i = 0; i < lines.length - 1; i++) {
+            var line = lines[i].split(delimiter);
+            result += '<tr> <td style="test-align: right">' + line[0];
+            result += '<td style="test-align: left">' + line[1];
```
`class="text-left"`

marusak replied:
Fixed

mvollmer commented:
Random comment: The small journal panels (on the Network and Storage page, for example) wont show the "Problem" entries, I guess.  This doesn't matter much right now since one can't do anything from those journal panels.  We can better with them in general.

andreasn commented:
> Random comment: The small journal panels (on the Network and Storage page, for example) wont show the "Problem" entries, I guess. This doesn't matter much right now since one can't do anything from those journal panels. We can better with them in general.

Does that mean that if I get a crash on udisksd or smartd, those won't show up on the logs of the individual pages?

mvollmer commented:
This should have a integration test, but for that we need to test on Fedora 26.  I think we can wait for that, since it's unlikely that the journal code will change a lot in the near future.

mvollmer commented:
> Does that mean that if I get a crash on udisksd or smartd, those won't show up on the logs of the individual pages?

Correct.

mvollmer commented:
Blocked on #6069

garrett commented:
@marusak: Sent! Check your inbox (@ redhat.com).

marusak commented:
> This should have a integration test, but for that we need to test on Fedora 26. I think we can wait for that, since it's unlikely that the journal code will change a lot in the near future.

Does this mean that this PR waits for Fedora 26 release? Or until you create testing on Fedora 26 (from alpha/beta)?

mvollmer commented:
> Or until you create testing on Fedora 26 (from alpha/beta)?

This.

petervo commented:
We have f26 testing now

marusak commented:
> We have f26 testing now

I know, however we need Beta version, since in Alpha some packages are missing.
Or, maybe, if your testing environment is not a pure Alpha, we can install missing packages there.

mvollmer commented:
> Or, maybe, if your testing environment is not a pure Alpha, we can install missing packages there.

We can do that, I guess, like we did with storaged pre-releases.  What do we need to install?  Can you make a PR that changes test/images/scripts/fedora-26.setup appropriately?

marusak commented:
> We can do that, I guess, like we did with storaged pre-releases. 

Happy to hear that!

> Can you make a PR that changes test/images/scripts/fedora-26.setup appropriately?

Sure. https://github.com/cockpit-project/cockpit/pull/6423

marusak commented:
So https://github.com/cockpit-project/cockpit/pull/6423 was merged. Could we continue on this?

mvollmer commented:
> Could we continue on this?

A integration test is missing.  I get this started, okay?

mvollmer commented on pkg/systemd/logs.js:
```diff
@@ -288,6 +380,391 @@ $(function() {
+        tab.append(d_btn);
+        tab.append(r_btn);
+
+        out.html(tab);
```
`out` is a `<table>` and you can't put things like a `<ul>` in there.  Maybe put it within a `<caption>`?  It's probably best to put a number of  `<table>`s into the `<li>`s, though.

mvollmer commented on pkg/systemd/logs.js:
```diff
@@ -288,6 +380,391 @@ $(function() {
+                if (elem in args){
+                    val = args[elem][2];
+                    c1.append(
+                        $('<tr>').append(
```
This puts a `<tr>` inside a `<div>`.  I don't think that's allowed.

mvollmer commented:
The layout breaks for small widths:

![screenshot from 2017-05-15 13-53-50](https://cloud.githubusercontent.com/assets/3228183/26054630/f74fe7de-3975-11e7-9f0f-b4171d01bca7.png)

The first table has different column width than the second (the one starting with "Directory"), and the first column in the first table is too narrow.

marusak commented:
> The first table has different column width than the second (the one starting with "Directory"), and the first column in the first table is too narrow.
I think it would be good enough to just have one long table in the "Problem info" tab. Otherwise we should probably use a Patternfly/Bootstrap grid to get proper responsive behavior.

Making it one column is much easier. If you think that is good enough.

garrett commented:
It's also possible to use a bit of CSS to force wrapping when there isn't space for it. 

Add `word-break: break-all;` to the appropriate value area in the CSS... and/or `whitespace: nowrap;` to the label CSS.

mvollmer commented:
I see wrong PIDs in the journal:

![screenshot from 2017-05-16 14-24-40](https://cloud.githubusercontent.com/assets/3228183/26103799/ce50a200-3a43-11e7-8b5a-86e145911ab6.png)

Note how systemd-coredump reports a different PID than abrt.  The abrt PID seems to come from the first crash of will_segfault.

This seems to be a bug in abrt itself, since the message in the journal is already wrong, not just how Cockpit renders it.

mvollmer commented on pkg/systemd/logs.js:
`lastr_occurence` and `time` are rendered as integers.  We should render them as date/time strings.

mvollmer commented on pkg/systemd/logs.js:
```diff
@@ -288,6 +380,391 @@ $(function() {
+        // write into general tab non-ABRT related items
+        var keys = Object.keys(entry).sort();
+        $.each(keys, function(i, key) {
+            if (key !== 'MESSAGE' && !key.startsWith('PROBLEM_')) {
```
`startsWith` is unfortunately not portable enough.  You could use `key.indexOf("PROBLEM_") == 0`.  Why do we filter out the PROBLEM_  fields, actually? I think it would be okay to keep them.

marusak replied:
Because it contains some ABRT related information, that are better displayed in Problem info. I found it redundant.

mvollmer commented on pkg/systemd/logs.js:
```diff
@@ -288,6 +380,391 @@ $(function() {
+    function create_problem(out, entry) {
+        var problem;
+        var all_p = [entry['PROBLEM_DIR'], entry['PROBLEM_DUPHASH'], entry['PROBLEM_UUID']];
+        for (var i = 0; i < 3; i++) {
```
`3`  -> `all_p.length`

mvollmer commented:
Here is a simple test: https://github.com/mvollmer/cockpit/commit/72c0fd8b82fa951b4c552488497996ec698bfc08

It currently fails because `displayable_problems` starts out as empty and never gets updated with the new problem, as indicated in my other review comments.

marusak commented on pkg/lib/journal.js:
```diff
@@ -230,7 +234,34 @@
+                }
+                //Do not render unknown (deleted) problems
+                if (! known)
+                    return "";
```
So you think that if the same problem happens more than once, we should display first, second and all other occurrences? It may however look a bit weird when you filter out only problems and there will be the same line several times.

I cannot agree with you about the count though. I think it is quite an important information. You will definitely pay more attention if you see that something crashed several times.

mvollmer replied:
> So you think that if the same problem happens more than once, we should display first, second and all other occurrences?

Yes.  The journal is a time-line of events.  We shouldn't mess with that by omitting entries.  Let's say something bad happened on your server around 14:23 o'clock.  You go to the journal to see what has happened around that time.  If abrt has written something to the log around that time you want to see it.  It should not be omitted because the same thing has also happened later at 16:15.

> I cannot agree with you about the count though. I think it is quite an important information. You will definitely pay more attention if you see that something crashed several times.

The list of journal entries is not the place to show this kind of summary, I'd say.  We can collect statistics from the journal and show them in a separate "Notifications" UI, in a very prominent place. We some old ideas about this here: https://github.com/cockpit-project/cockpit/wiki/Notifications, maybe abrt is the thing that can drive this.

marusak replied:
> Yes. The journal is a time-line of events....

Yeah, this makes a better sense.

> The list of journal entries is not the place to show this kind of summary...

Ok than. The design specified that this should dispaly the problem count and I liked that idea. However, if you believe it is not a good feature, we can omit it easily.

> ...  "Notifications" UI ...

I like this! When it gets implemented one day, it would be useful to inform about ABRT problems there.

andreasn replied:
In the design, I intended the counter only for things happening right after each other (to reduce noise from services that fails constantly and spits out lines), so it's consistent with how the rest of the logs page work. Sorry if the design was unclear there.

For situations where a crash happens, say, once a day, I feel it's important that you see that it's reoccurring on a predictable time-basis, and that the entry's for different days don't get collapsed down.

I agree with the idea of the notification area. It will be needed also for things like software updates being available and other critical things, so let's solve that as part of that.

marusak commented on pkg/systemd/logs.js:
It is not. However if you decide that that we should display all logs - first, second, nth occurrence of the same problem, than we can drop this - probably. One of the purposes of this was to know the count of each problem so we could only render the last log.

marusak commented:
> Here is a simple test: mvollmer/cockpit@72c0fd8

Thanks!

I am going to fix all of yours comments and then try this test. If it works, than I may try to add another tests.

However it may take a while since my final state examination is getting closer and I need to study for this.

mvollmer commented:
> I am going to fix all of yours comments and then try this test.

Great!  I have a pretty concrete idea of the changes that I would make, so I might work a bit on this while you study. :-)

Thanks a lot for the work so far!

marusak commented:
> Great! I have a pretty concrete idea of the changes that I would make, so I might work a bit on this while you study. :-)

That would be awesome:) If you have anything, feel free to force push anytime.

Thank you for the reviews and help!

mvollmer commented on pkg/systemd/logs.js:
```diff
@@ -28,6 +28,71 @@ $(function() {
     cockpit.translate();
     var _ = cockpit.gettext;
 
+    var problems_client = cockpit.dbus('org.freedesktop.problems');
```
This should use `{ "superuser": "try" }` to match how journalctl is run.  If possible, this will then talk to the daemon as root and be able to see more problems.

marusak commented:
@mvollmer I finally got to this. I fixed all the issues you wrote above. I also extended your integration test. All works well but one test does not wait long enough and I cannot make it wait longer...
https://github.com/cockpit-project/cockpit/pull/6189/commits/855ee91c1b80165b4b1c3966ea23709640b8cae0#diff-b06317782727dc6a2826cc478f09bb06R362 in this part the code changes location (  https://github.com/cockpit-project/cockpit/pull/6189/commits/a6854571149fd39c95b1272d5107446e77dc4925#diff-b6e3e8699e92d8119693dabf74886a96R438   ) and I tried pretty much everything but without any success. Is there something that I am missing?

marusak commented:
I finally got back to this. Fixed all issues that @mvollmer mentioned above.
I also wrote integration tests.
@mvollmer what do you think of this now?

martinpitt commented:
I triggered a run of the integration tests.

mvollmer commented:
> @mvollmer what do you think of this now?

Very nice!

martinpitt commented:
Some screenshots for the release notes:

![abrt-list](https://user-images.githubusercontent.com/200109/27947072-b6de7726-62f4-11e7-875b-1ad9f877e2d1.png)

![abrt-details](https://user-images.githubusercontent.com/200109/27947075-b9db4f30-62f4-11e7-8ecf-dcaff37029d3.png)
